File: src/processOrderVanilla.ts
import { orders, books } from './data'
import { validateOrder, Order, AsyncProcessor, PlacedOrderResult } from './api'

const bookService = (bookId: string) =>
  books[bookId]
    ? Promise.resolve(books[bookId])
    : Promise.reject(new Error(`Book not found: ${bookId}`))

const orderService = (orderId: string) =>
  orders[orderId]
    ? Promise.resolve(orders[orderId])
    : Promise.reject(new Error(`Order not found: ${orderId}`))
const validationService = (order: Order) => {
  const r = validateOrder(order)
  if (r.valid) {
    return Promise.resolve(order)
  } else {
    return Promise.reject(new Error(`${r.error}`))
  }
}

const calculateAmountService = async (order: Order) => {
  let total = 0
  for (let i = 0; i < order.items.length; i++) {
    const item = order.items[i]
    const book = await bookService(item.bookId)
    total += item.quantity * book.price
  }
  return total
}

const placeOrderService = (order: Order) =>
  calculateAmountService(order).then(
    totalAmount =>
      ({
        success: true,
        totalAmount
      } as PlacedOrderResult)
  )

const processor: AsyncProcessor = (
  orderId: string
): Promise<PlacedOrderResult> => {
  return orderService(orderId)
    .then(validationService)
    .then(placeOrderService)
    .catch(() => {
      return {
        success: false,
        totalAmount: 0.0
      }
    })
}

export default processor

File: rust/src/process_order_vanilla.rs
use crate::api::*;
use crate::data::{get_book, get_order};
use async_trait::async_trait;

async fn book_service(id: &String) -> Option<&'static Book> {
    get_book(id)
}

async fn order_service(id: &String) -> Option<&'static Order> {
    get_order(id)
}

async fn validation_service(order: &Order) -> Result<&Order, OrderNotValid> {
    match validate_order(order) {
        Ok(_) => Ok(order),
        Err(err) => Err(err),
    }
}

async fn calculate_amount_service(order: &Order) -> (f64, &Order) {
    let mut total = 0.0;
    for item in &order.items {
        let book = book_service(&item.book_id).await;
        match book {
            Some(b) => {
                total += item.quantity as f64 * b.price;
            }
            _ => {}
        };
    }
    (total, order)
}

async fn place_order_service(order: &Order) -> PlacedOrderResult {
    let result = calculate_amount_service(order).await;
    Ok(OrderSuccessful::new(result.0))
}

pub struct VanillaProcessor {}

#[async_trait]
impl AsyncProcessor for VanillaProcessor {
    async fn process(&self, order_id: &String) -> f64 {
        let order = order_service(order_id).await;
        match order {
            Some(order) => {
                let validation = validation_service(&order).await;
                match validation {
                    Ok(_) => match place_order_service(order).await {
                        Ok(res) => res.amount,
                        Err(_) => 0.0,
                    },
                    _ => 0.0,
                }
            }
            _ => 0.0,
        }
    }
}

impl VanillaProcessor {
    pub fn processor() -> &'static dyn AsyncProcessor {
        &(VanillaProcessor {}) as &dyn AsyncProcessor
    }
}
